src/librustc_mir/transform/instrument_coverage.rs

```diff
@@ -79,14 +126,59 @@ fn function_handle<'tcx>(tcx: TyCtxt<'tcx>, fn_def_id: DefId, span: Span) -> Ope
     Operand::function_handle(tcx, fn_def_id, substs, span)
 }
 
-fn placeholder_block<'tcx>(source_info: SourceInfo) -> BasicBlockData<'tcx> {
+fn placeholder_block(span: Span) -> BasicBlockData<'tcx> {
     BasicBlockData {
         statements: vec![],
         terminator: Some(Terminator {
-            source_info,
+            source_info: SourceInfo::outermost(span),
             // this gets overwritten by the counter Call
             kind: TerminatorKind::Unreachable,
         }),
         is_cleanup: false,
     }
 }
+
+fn hash_mir_source<'tcx>(tcx: TyCtxt<'tcx>, src: &MirSource<'tcx>) -> u64 {
+    let fn_body_id = match tcx.hir().get_if_local(src.def_id()) {
+        Some(node) => match associated_body(node) {
+            Some(body_id) => body_id,
+            _ => bug!("instrumented MirSource does not include a function body: {:?}", node),
+        },
+        None => bug!("instrumented MirSource is not local: {:?}", src),
+    };
+    let hir_body = tcx.hir().body(fn_body_id);
+    let mut hcx = tcx.create_no_span_stable_hashing_context();
+    hash(&mut hcx, &hir_body.value).to_smaller_hash()
+}
+
+fn hash(
+    hcx: &mut StableHashingContext<'tcx>,
+    node: &impl HashStable<StableHashingContext<'tcx>>,
+) -> Fingerprint {
+    let mut stable_hasher = StableHasher::new();
+    node.hash_stable(hcx, &mut stable_hasher);
+    stable_hasher.finish()
+}
+
+fn associated_body<'hir>(node: Node<'hir>) -> Option<BodyId> {
```

Comment:
FYI, this was copied verbatim from:

https://github.com/richkadel/rust/blob/754da8849c18c45ff2fd2e77213f8371de488f80/src/librustc_middle/hir/map/mod.rs#L59

If there's no objection, I'd like to make the original function public and just use it directly.

Reply:
Making it public seems fine. I'm not sure it should (continue to) live there, but I'm also not sure where it should go instead.

Reply:
done